sdk/src/main/java/co/omise/android/ui/EContextFormFragment.kt

```diff
@@ -0,0 +1,125 @@
+    private val phoneNumberErrorText by lazy { text_phone_number_error }
+    private val submitButton: Button by lazy { button_submit }
+    private val formInputWithErrorTexts: List<Pair<OmiseEditText, TextView>> by lazy {
+        listOf(
+                Pair(fullNameEdit, fullNameErrorText),
+                Pair(emailEdit, emailErrorText),
+                Pair(phoneNumberEdit, phoneNumberErrorText)
+        )
+    }
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        return inflater.inflate(R.layout.fragment_econtext_form, container, false)
+    }
+
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+
+        title = when (type) {
+            SupportedEContext.ConvenienceStore -> getString(R.string.title_convenience_store)
+            SupportedEContext.PayEasy -> getString(R.string.title_pay_easy)
+            SupportedEContext.Netbanking -> getString(R.string.title_netbanking)
+            null -> getString(R.string.econtext_title)
+        }
+        setHasOptionsMenu(true)
+
+        formInputWithErrorTexts.forEach {
+            it.first.setOnFocusChangeListener(::updateErrorText)
+            it.first.setOnAfterTextChangeListener(::updateSubmitButton)
+        }
+
+        submitButton.setOnClickListener(::submitForm)
+    }
+
+    private fun updateErrorText(view: View, hasFocus: Boolean) {
+        val editText = view as OmiseEditText
+        val errorText = formInputWithErrorTexts.first { it.first == editText }.second
+
+        if (hasFocus || editText.isValid) {
+            errorText.text = ""
+            errorText.visibility = INVISIBLE
+            return
+        }
+
+        val errorMessage = when (editText) {
+            fullNameEdit -> R.string.error_invalid_full_name
+            emailEdit -> R.string.error_invalid_email
+            phoneNumberEdit -> R.string.error_invalid_phone_number
+            else -> R.string.error_invalid_unknown
+        }
+
+        errorText.text = getString(errorMessage)
+        errorText.visibility = VISIBLE
+    }
+
+    private fun updateSubmitButton() {
+        val isFormValid = formInputWithErrorTexts.map { it.first.isValid }
+                .reduce { acc, b -> acc && b }
+        submitButton.isEnabled = isFormValid
+    }
+
+    private fun submitForm() {
+        val requester = requester ?: return
+
+        val fullName = fullNameEdit.text?.toString()?.trim().orEmpty()
+        val email = emailEdit.text?.toString()?.trim().orEmpty()
+        val phoneNumber = phoneNumberEdit.text?.toString()?.trim().orEmpty()
+
+        val request = Source.CreateSourceRequestBuilder(requester.amount, requester.currency, SourceType.Econtext)
+                .name(fullName)
+                .email(email)
+                .phoneNumber(phoneNumber)
+                .build()
+
+        view?.let { setAllViewsEnabled(it, false) }
```

Comment:
Can the `let` be replace by a direct call on `view?`?

Reply:
Sorry, `setAllViewsEnabled()` couldn't pass the nullable `view`. 😄